### PR TITLE
Cleaning TensorFlow models

### PR DESCRIPTION
**Pull request by LysandreJik**
While writing docstrings for #5036, I stumbled upon a few bugs in TensorFlow, especially related to the loss computation. I'm patching them in this PR.

Here's the list of the bugs solved:

### Loss computation

- The loss is computed differently than it is with the PyTorch models. Here it is returned example-wise (therefore with a shape of `(batch_size,)`, whereas the PyTorch models return the loss as a scalar. @jplu is there a reason for this implementation?
- The TensorFlow models should be able to handle three types of scenarios: keyword arguments, dictionary, and tuple/list. Right now the `labels` can only be passed through the keyword argument. This PR changes that, and adds a test.

### Missing models in the test files

A few models were implemented but were not tested. Some of these models were not working as expected, therefore they've been updated.

- TF DistilBERT for multiple choice (added test and patched)
- TF DistilBERT for token classification
- TF Electra for QA
- TF RoBERTa for multiple choice
- TF XLNet for multiple choice (added test and patched)

### Misc

- Most of the QA models had `is_impossible`, `cls_index` and `p_mask` in their signature while not making use of them. These have been removed. **Users relying on the order of arguments in the signature will be affected by this**
- The `labels` were generally placed before the `output_attentions` and `output_hidden_states` that have recently been added to the models. This resulted in an error in the documentation as the `labels` (part of the head model) were added after the `output_attentions` and `output_hidden_states` (part of the base model). The arguments have been re-ordered to once again respect the order of `**base_arguments, **head_arguments`

### CODE REVIEW

**Review by sgugger — APPROVED** [review]
This is great work to clean-up some bugs in non-tested models! Thanks for doing the TF side @LysandreJik 

Side-note: should we have some kind of test run in `make quality` that actually checks all models are properly tested? This way it would flag in red a PR that adds a model but doesn't properly test it, which would avoid having to manually check everything once in a while like in this PR or the equivalent for PyTorch models last week.

**Comment by jplu** [comment]
Thanks @LysandreJik !

> The loss is computed differently than it is with the PyTorch models. Here it is returned example-wise (therefore with a shape of (batch_size,), whereas the PyTorch models return the loss as a scalar. @jplu is there a reason for this implementation?

Because it is the generic approach to use as some of the other reductions are not compliant with custom training loop. For example I see that you have used `SUM_OVER_BATCH_SIZE` instead of `None` but this removes the compatibility with custom training loops like we have in the trainer, see the [doc](https://www.tensorflow.org/api_docs/python/tf/keras/losses/Reduction). Then can you undo this part please.

I do the reduction then directly in the trainer and not in the model, but we can do the reduction manually inside either the loss functions, or the `call` methods as you wish :)

> The TensorFlow models should be able to handle three types of scenarios: keyword arguments, dictionary, and tuple/list. Right now the labels can only be passed through the keyword argument. This PR changes that, and adds a test.

Good catch, thanks for having fixed this!

> Most of the QA models had is_impossible, cls_index and p_mask in their signature while not making use of them. These have been removed. Users relying on the order of arguments in the signature will be affected by this

Ok, I didn't know, when I reworked the TF models, I mostly took examples on the list of parameters from the PT part at a time T, I should have been more carefull on later changes. Sorry.

> The labels were generally placed before the output_attentions and output_hidden_states that have recently been added to the models. This resulted in an error in the documentation as the labels (part of the head model) were added after the output_attentions and output_hidden_states (part of the base model). The arguments have been re-ordered to once again respect the order of **base_arguments, **head_arguments

Thanks!! Like previously I should have been more careful on recent changes. My bad.

File: src/transformers/modeling_tf_utils.py [hunk 1]
```diff
@@ -109,7 +109,7 @@ def get_config(self):
 class TFQuestionAnsweringLoss:
     def compute_loss(self, labels, logits):
         loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(
-            from_logits=True, reduction=tf.keras.losses.Reduction.NONE
+            from_logits=True, reduction=tf.keras.losses.Reduction.SUM_OVER_BATCH_SIZE
```
Comment (jplu) [inline src/transformers/modeling_tf_utils.py:112]: NONE should be kept here for compatibility reasons. (same for the others below)

**Comment by LysandreJik** [comment]
Okay, thanks for the review @jplu, I'll revert that part.

**Comment by LysandreJik** [comment]
I'll update the documentation in the next PR

**Comment by patrickvonplaten** [comment]
That's awesome, thanks @LysandreJik !

**Review by thomwolf — APPROVED** [review]
